### PR TITLE
 [PIP] Simple data loader

### PR DESCRIPTION
**Pull request by karatakis**
## PR Info
Allows to query related entities from vector

### Load one
```rust
[...]
    let bakers = baker::Entity::find()
        .all(&ctx.db)
        .await
        .expect("Should load bakers");

    let bakeries = bakers
        .load_one(bakery::Entity::find(), &ctx.db)
        .await
        .expect("Should load bakeries");
[...]
```

### Load many

```rust
[...]
    let bakeries = bakery::Entity::find()
        .all(&ctx.db)
        .await
        .expect("Should load bakeries");

    let bakers = bakeries
        .load_many(baker::Entity::find().filter(baker::Column::Name.like("Baker%")), &ctx.db)
        .await
        .expect("Should load bakers");
[...]
```

- Closes https://github.com/SeaQL/sea-orm/issues/1191

## New Features

- [X] Add simple Dataloader trait

### CODE REVIEW

**Comment by karatakis** [comment]
Its not ready yet. I need to do some testing

**Comment by tyt2y3** [comment]
The implementation looks great!

Made a few improvements:
1. Use ValueTuple to replace Vec<Value>
2. Remove Debug trait bounds
3. Slightly improve the test case

May be we still want more test cases?

I am tempted to change `HashMap<String, _>` to `HashMap<ValueTuple, _>`, but of course, `Value` is not hashable because of f32/f64. And I imagine the attempt to make a hashable version of Value to be quite difficult.

File: src/query/loader.rs [hunk 1]
```diff
@@ -0,0 +1,340 @@
...
+                .in_tuples(keys.iter().cloned()),
+            )
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    #[tokio::test]
+
```
Comment (tyt2y3) [inline src/query/loader.rs:255]: How do I remove this line?
```suggestion

```

File: src/query/loader.rs [hunk 2]
```diff
@@ -0,0 +1,340 @@
...
+            cakes,
+            vec![Some(cake::Model {
+                id: 1,
+                name: "New York Cheese".to_owned(),
+            })]
+        );
+    }
+
+    #[tokio::test]
+
```
Comment (tyt2y3) [inline src/query/loader.rs:297]:
```suggestion

```

**Comment by karatakis** [comment]
@tyt2y3 thanks for the feedback and the extra work.

@tyt2y3 do you want me to work on making Value Hash 'able?
If you do assign an issue and assign it to me to work on it :smile:.
This is also needed in some cases for Seaography.

Is anything else needed for this issue ?

**Comment by tyt2y3** [comment]
Yes we can come back to this after we got a `HashableValue` .

Also may be awaiting @billy1624 's review

**Comment by tyt2y3** [comment]
> 2.1. we should assume that the related entity might be shared; i.e. A1 & A2 both relates to B1, we will only get one B1, so we have to clone it

Can we also test this?

**Comment by tyt2y3** [comment]
Created issue for HashableValue
https://github.com/SeaQL/sea-orm/issues/1254

**Comment by karatakis** [comment]
Regarding https://github.com/SeaQL/sea-orm/pull/1238#issuecomment-1326635798

It didn't panic, but it was returning None :'(. I fixed it by copying the values before returning.

**Review by billy1624 — COMMENTED** [review]
Hey @karatakis, sorry for coming in late. nice stuffs!! Thanks for the hard work!!

I think we should also implement the loader for `M: ModelTrait`. Providing a "complete" API, `Vec<M>` implemented `LoaderTrait`, `M` itself should also implement it.

#### Load one

```rust
let bakers = baker::Entity::find()
    .one(&ctx.db) // `one` here, instead of all
    .await?;

let bakeries: Option<_> = bakers // `Option<_>` here, instead of `Vec<Option<_>>`
    .load_one(bakery::Entity::find(), &ctx.db)
    .await?;
```

#### Load many

```rust
let bakeries = bakery::Entity::find()
    .one(&ctx.db) // `one` here, instead of all
    .await?;

let bakers: Vec<_> = bakeries // `Vec<_>` here, instead of `Vec<Vec<_>>`
    .load_many(baker::Entity::find().filter(baker::Column::Name.like("Baker%")), &ctx.db)
    .await?;
```

That means we need two extra associate types to type def the return type of both methods.

```rust
pub trait LoaderTrait {
    // Existing associated type
    type Model: ModelTrait;

    // New associated type to type def the return types
    type LoadOneResult;
    type LoadManyResult;
```

**Comment by tyt2y3** [comment]
> I think we should also implement the loader for M: ModelTrait. Providing a "complete" API, Vec<M> implemented LoaderTrait, M itself should also implement it.

Good point, although I think we don't have to mess the implementation, I suppose we can simply provide an implementation that just wraps a Vec of 1 item.

I want to merge this for 0.11

File: src/query/loader.rs [hunk 3]
```diff
@@ -0,0 +1,339 @@
...
+    ///  Used to eager load has_many relations
+    async fn load_many<R, C>(&self, stmt: Select<R>, db: &C) -> Result<Vec<Vec<R::Model>>, DbErr>
+    where
+        C: ConnectionTrait,
+        R: EntityTrait,
+        R::Model: Send + Sync,
+        <<Self as LoaderTrait>::Model as ModelTrait>::Entity: Related<R>;
+}
+
+#[async_trait::async_trait]
```
Comment (tyt2y3) [inline src/query/loader.rs:32]:
```suggestion
#[async_trait]
```

**Pull request merged**